app/main/views/suppliers.py

```diff
@@ -471,3 +481,80 @@ def update_supplier_framework_details(supplier_id, framework_slug):
         return jsonify(message="Database Error: {0}".format(e)), 400
 
     return jsonify(frameworkInterest=interest_record.serialize()), 200
+
+
+@main.route('/suppliers/<int:supplier_id>/frameworks/<framework_slug>/variation/<variation_slug>', methods=['PUT'])
+def agree_framework_variation(supplier_id, framework_slug, variation_slug):
+
+    framework = Framework.query.filter(
+        Framework.slug == framework_slug
+    ).first_or_404()
+
+    supplier = Supplier.query.filter(
+        Supplier.supplier_id == supplier_id
+    ).first_or_404()
+
+    # TODO SELECT FOR UPDATE?
+    # (technically we should be doing a FOR UPDATE of SupplierFramework and a FOR SHARE of Framework)
+    interest_record = SupplierFramework.query.filter(
+        SupplierFramework.supplier_id == supplier.supplier_id,
+        SupplierFramework.framework_id == framework.id
+    ).first()
+    if not interest_record:
+        abort(404, "supplier_id '{}' has not registered interest in {}".format(supplier_id, framework_slug))
+
+    if not interest_record.on_framework:
+        abort(404, "supplier_id '{}' is not on framework {}".format(supplier_id, framework_slug))
+
+    # acceptable variation_slugs must have corresponding createdAt entry in variations dict
+    if not (framework.framework_agreement_details or {}).get("variations", {}).get(variation_slug, {}).get("createdAt"):
+        abort(404, "Unknown variation '{}' for framework {}".format(variation_slug, framework_slug))
+
+    if (interest_record.agreed_variations or {}).get(variation_slug, {}).get("agreedAt"):
```

Comment:
I'm not sure this should be an API error here - could it just return the existing `agreedVariation` details? You're generally supposed to be able to PUT multiple times to the same thing I think...